src/jit/codegenxarch.cpp

```diff
@@ -8396,6 +8366,84 @@ void CodeGen::genSSE2BitwiseOp(GenTreePtr treeNode)
     inst_RV_RV(ins, targetReg, operandReg, targetType);
 }
 
+#ifdef _TARGET_X86_
+void CodeGen::genFld(GenTree* source)
+{
+    assert(source != nullptr);
+    assert(!source->isContained());
+
+    // NOTE: the caller must call genConsumReg for the source.
+
+    // Spill the return value register from an XMM register to the stack, then load it on the x87 stack.
+    // If it already has a home location, use that. Otherwise, we need a temp.
+    if (genIsRegCandidateLocal(source) && compiler->lvaTable[source->gtLclVarCommon.gtLclNum].lvOnFrame)
+    {
+        // Store local variable to its home location, if necessary.
+        if ((source->gtFlags & GTF_REG_VAL) != 0)
+        {
+            source->gtFlags &= ~GTF_REG_VAL;
+            inst_TT_RV(ins_Store(source->gtType,
+                                 compiler->isSIMDTypeLocalAligned(source->gtLclVarCommon.gtLclNum)),
+                       source, source->gtRegNum);
+        }
+        // Now, load it to the fp stack.
+        getEmitter()->emitIns_S(INS_fld, emitTypeSize(source), source->AsLclVarCommon()->gtLclNum, 0);
+    }
+    else
+    {
+        // Spill the value, which should be in a register, then load it to the fp stack.
+        // TODO-X86-CQ: Deal with things that are already in memory (don't call genConsumeReg yet).
+        source->gtFlags |= GTF_SPILL;
+        regSet.rsSpillTree(source->gtRegNum, source);
+        assert((source->gtFlags & (GTF_SPILLED | GTF_SPILL)) == GTF_SPILLED);
+
+        TempDsc* t = regSet.rsUnspillInPlace(source, source->gtRegNum);
+        inst_FS_ST(INS_fld, emitActualTypeSize(source->gtType), t, 0);
+        source->gtFlags &= ~GTF_SPILLED;
+        compiler->tmpRlsTemp(t);
+    }
+}
+
+void CodeGen::genX87Intrinsic(GenTreeIntrinsic* node)
+{
+    assert(node != nullptr);
+    assert(node->TypeGet() == TYP_DOUBLE);
+
+    instruction instr;
+    switch (node->gtIntrinsicId)
+    {
+        case CORINFO_INTRINSIC_Cos:
+            instr = INS_fcos;
+            break;
+
+        case CORINFO_INTRINSIC_Sin:
+            instr = INS_fsin;
+            break;
+
+        case CORINFO_INTRINSIC_Round:
+            instr = INS_frndint;
+            break;
+
+        default:
+            unreached();
+    }
+
+    GenTree* source = node->gtOp1;
+    assert(source->TypeGet() == TYP_DOUBLE);
+
+    // Load the source onto the x87 stack.
+    genConsumeReg(source);
+    genFld(source);
```

Comment:
Here is a clean and easy way to achieve what genFld() does plus it also allows us to optimize for the case when the operand is a memory-op:

- Compiler class declares two members to hold local numbers, these are initialized to BAD_VAR_NUM
floatTemp and doubleTemp
- TreeNodeInfoInitIntrinsic() - in case of cos/sin/round intrinsics:
if the operand is a memory-op, mark it as contained, otherwise create a temp of float/double depending on target type of GT_INTRINSIC node.
```
e.g.  if (floatTemp == BAD_VAR_NUM)
       {
            floatTemp = comp->lvaGrabTemp(...)
            // initialize lv type, lvRefCnt, lvDoNotEnregister
            // set lvaSortAgain to true
       }
```
- genIntrinsic() - in case of sin/cos/round, if the operand is not a contained memor-op, use floatTemp/doubleTemp to store the value of the operand to memory and to copy to x87 stack from memory.  If it is a contained memory op, we can use it directly.